roles/openshift_health_checker/openshift_checks/diagnostics.py

```diff
@@ -0,0 +1,62 @@
+"""
+A check to run relevant diagnostics via `oc adm diagnostics`.
+"""
+
+import os
+
+from openshift_checks import OpenShiftCheck, OpenShiftCheckException
+
+
+DIAGNOSTIC_LIST = (
+    "AggregatedLogging ClusterRegistry ClusterRoleBindings ClusterRoles "
+    "ClusterRouter DiagnosticPod NetworkCheck"
+).split()
+
+
+class DiagnosticCheck(OpenShiftCheck):
+    """A check to run relevant diagnostics via `oc adm diagnostics`."""
+
+    name = "diagnostics"
+    tags = ["health"]
+
+    def is_active(self):
+        return super(DiagnosticCheck, self).is_active() and self.is_first_master()
+
+    def run(self):
+        if self.exec_diagnostic("ConfigContexts"):
+            # only run the other diagnostics if that one succeeds (otherwise, all will fail)
+            diagnostics = self.get_var("openshift_check_diagnostics", default=DIAGNOSTIC_LIST)
+            for diagnostic in self.normalize(diagnostics):
+                self.exec_diagnostic(diagnostic)
+        return {}
+
+    def exec_diagnostic(self, diagnostic):
+        """
+        Execute an 'oc adm diagnostics' command on the remote host.
+        Raises OcNotFound or registers OcDiagFailed.
+        Returns True on success or False on failure (non-zero rc).
+        """
+        config_base = self.get_var("openshift.common.config_base")
+        args = {
+            "config_file": os.path.join(config_base, "master", "admin.kubeconfig"),
+            "cmd": "adm diagnostics",
+            "extra_args": [diagnostic],
+        }
+
+        result = self.execute_module("ocutil", args, save_as_name=diagnostic + ".failure.json")
+        self.register_file(diagnostic + ".txt", result['result'])
```

Comment:
Nope; that's the output from `oc adm diagnostics` and it's not JSON.

However... now that I look at it... the `execute_module` `save_as_name` shouldn't have `failure` in the name. It's just saving whatever the module result hash was, not necessarily failed.

Reply:
I was confused because I saw you wrote to `diagnostic + ".failure.json"` but did not use it afterwards; are these files written so that people who know to look for them can find them?

Reply:
They're written in an output directory if provided.

https://github.com/openshift/openshift-ansible/blob/d09223f529e37fd4983bbb29f27edb403bbe010d/roles/openshift_health_checker/openshift_checks/__init__.py#L133-L18

Reply:
Is there a need to save both the full json file and the result text file?

Reply:
Multiline text strings are difficult to read in JSON. Also frankly they are unlikely to care about the rest of the module result... I can consider whether it needs to be registered at all.